File: 0x08-python-more_classes/4-rectangle.py
#!/usr/bin/python3
"""
    Define class Rectangle
"""


class Rectangle:
    """
        This represente a Rectangle
    """

    def __init__(self, w=0, h=0):
        """
            Initialize a new Rectangle
            Args:
                h (int): Rectangle height
                w (int): Rectangle width
        """
        self.__height = h
        self.__width = w

    def _getWidth(self):
        """
            Width Getter
        """
        return self.__width

    def _setWidth(self, newWidth):
        """
            Width Setter
        """
        if not isinstance(newWidth, int):
            raise TypeError("width must be an integer")
        if newWidth < 0:
            raise ValueError("width must be >= 0")
        self.__width = newWidth

    def _getHeight(self):
        """
            Height Getter
        """
        return self.__height

    def _setHeight(self, newHeight):
        """
            Height Setter
        """
        if not isinstance(newHeight, int):
            raise TypeError("height must be an integer")
        if newHeight < 0:
            raise ValueError("height must be >= 0")
        self.__height = newHeight

    def area(self):
        """
            Return the area of the Rectangle
        """
        return (self.__width * self.__height)

    def perimeter(self):
        """
            Return the Rectangle perimeter
        """
        if self.__width == 0 or self.__height == 0:
            return 0
        return ((self.__width * 2) + (self.__height * 2))

    def __str__(self):
        """
            Print the Rectangle width #
        """
        if self.__width == 0 or self.__height == 0:
            return ""

        rectangle = []
        for i in range(self.__height):
            [rectangle.append('#') for j in range(self.__width)]
            if i != self.__height - 1:
                rectangle.append("\n")
        return "".join(rectangle)

    def __repr__(self):
        representation = ""
        representation += "Rectangle("
        representation += str(self.__width)
        representation += ", "
        representation += str(self.__height)
        representation += ")"
        return representation

    width = property(_getWidth, _setWidth)
    height = property(_getHeight, _setHeight)

File: 0x13-javascript_objects_scopes_closures/4-rectangle.js
#!/usr/bin/node

class Rectangle {
  constructor (w, h) {
    if (w > 0 && h > 0) {
      this.width = w;
      this.height = h;
    }
  }

  print () {
    for (let y = 0; y < this.height; y++) {
      let line = '';
      for (let x = 0; x < this.width; x++) {
        line += 'X';
      }
      console.log(line);
    }
  }

  rotate () {
    const w = this.width;
    const h = this.height;

    this.width = h;
    this.height = w;
  }

  double () {
    const doubleWidth = this.width * 2;
    const doubleHeight = this.height * 2;

    this.width = doubleWidth;
    this.height = doubleHeight;
  }
}

module.exports = Rectangle;
